File: GanServer/src/com/xattacker/json/JsonObject.java
package com.xattacker.json;

import com.google.gson.GsonBuilder;

public abstract class JsonObject implements JsonSerializable, JsonBuilderVisitor
{
	public void onBuilderPrepared(GsonBuilder aBuilder)
	{	
	}
	
	@Override
	public String toJson()
	{
		return JsonUtility.createGson(this).toJson(this);
	}
}

File: GanAndroidApp/GanClient/src/main/java/com/xattacker/json/JsonObject.kt
package com.xattacker.json

import com.google.gson.GsonBuilder
import com.xattacker.gan.data.FunctionType
import com.xattacker.gan.data.FunctionTypeSerializer

abstract class JsonObject : JsonSerializable
{
    open fun onBuilderPrepared(builder: GsonBuilder)
    {
    }

    override fun toJson(): String
    {
        return JsonUtility.createGson {
            builder: GsonBuilder ->
            onBuilderPrepared(builder)
        }.toJson(this)
    }
}